lib/tlscontext.c

```diff
@@ -697,6 +675,50 @@ tls_context_load_key_and_cert(TLSContext *self)
   return TLS_CONTEXT_OK;
 }
 
+static void _write_line_to_keylog_file_mutex(const char *file_path, const char *line)
+{
+  FILE *keylog_file;
+  gint ret_val;
+  static GStaticMutex mutex;
+  g_static_mutex_lock(&mutex);
+  keylog_file = fopen(file_path, "a");
+  if(keylog_file)
+    {
+      ret_val = fprintf(keylog_file, "%s\n", line);
+      if (ret_val != strlen(line))
+        {
+          msg_error("Couldn't write to TLS keylogfile", evt_tag_errno("error", ret_val));
+        }
+      fclose(keylog_file);
+    }
+  g_static_mutex_unlock(&mutex);
+}
+
+static inline void
+_dump_tls_keylog(const SSL *ssl, const char *line)
+{
+  if(ssl == NULL)
+    {
+      return;
+    }
+  const char *file_path = SSL_CTX_get_ex_data(SSL_get_SSL_CTX(ssl), _TLS_KEYLOG_INDEX);
+  _write_line_to_keylog_file_mutex(file_path, line);
+}
+
+static inline void
+_set_keylog_to_file(TLSContext *self)
+{
+#if OPENSSL_VERSION_NUMBER >= 0x10101000L
+  {
+    SSL_CTX_set_ex_data(self->ssl_ctx, _TLS_KEYLOG_INDEX, self->keylog_file);
+    SSL_CTX_set_keylog_callback(self->ssl_ctx, _dump_tls_keylog);
+  }
+#else
+  msg_error("SSL_CTX_set_keylog_callback function is not available in your OpenSSL version, please upgrade it to at least 1.1.1",
+            evt_tag_str("Your version", OPENSSL_VERSION_NUMBER));
+#endif
```

Comment:
Ifdefed compatibility implementations should go into `lib/compat/openssl_support`.

Since this is not an alternative implementation, just error reporting, I'd report this error sooner.
Why don't we add a `boolean` return value to `tls_context_set_keylog_file()`? That way, we could report the missing functionality as a config parsing error with `CHECK_ERROR()`.